pkg/v1/layout/write.go

```diff
@@ -221,34 +224,91 @@ func (l Path) WriteFile(name string, data []byte, perm os.FileMode) error {
 // WriteBlob copies a file to the blobs/ directory in the Path from the given ReadCloser at
 // blobs/{hash.Algorithm}/{hash.Hex}.
 func (l Path) WriteBlob(hash v1.Hash, r io.ReadCloser) error {
+	return l.writeBlob(hash, -1, r, nil)
+}
+
+func (l Path) writeBlob(hash v1.Hash, size int64, r io.Reader, renamer func() (v1.Hash, error)) error {
+	if hash.Hex == "" && renamer == nil {
+		panic("writeBlob called an invalid hash and no renamer")
+	}
+
 	dir := l.path("blobs", hash.Algorithm)
 	if err := os.MkdirAll(dir, os.ModePerm); err != nil && !os.IsExist(err) {
 		return err
 	}
 
+	// Check if blob already exists and is the correct size
 	file := filepath.Join(dir, hash.Hex)
-	if _, err := os.Stat(file); err == nil {
-		// Blob already exists, that's fine.
+	if s, err := os.Stat(file); err == nil && !s.IsDir() && (s.Size() == size || size == -1) {
 		return nil
 	}
-	w, err := os.Create(file)
+
+	// If a renamer func was provided write to a temporary file
+	open := func() (*os.File, error) { return os.Create(file) }
+	if renamer != nil {
+		open = func() (*os.File, error) { return ioutil.TempFile(dir, hash.Hex) }
+	}
+	w, err := open()
 	if err != nil {
 		return err
 	}
+	if renamer != nil {
+		// Delete temp file if an error is encountered before renaming
+		defer os.Remove(w.Name())
+	}
 	defer w.Close()
 
-	_, err = io.Copy(w, r)
-	return err
-}
+	// Write to file and exit if not renaming
+	if n, err := io.Copy(w, r); err != nil || renamer == nil {
+		return err
+	} else if size != -1 && n != size {
+		return fmt.Errorf("expected blob size %d, but only wrote %d", size, n)
+	}
 
-// TODO: A streaming version of WriteBlob so we don't have to know the hash
-// before we write it.
+	// Always close file before renaming
+	if err := w.Close(); err != nil {
+		return err
+	}
 
-// TODO: For streaming layers we should write to a tmp file then Rename to the
-// final digest.
+	// Rename file based on the final hash
+	finalHash, err := renamer()
```

Comment:
Does this need an `if renamer != nil` ?

Reply:
No, because the conditional on line 262 causes a return if renamer is nil: `if n, err := io.Copy(w, r); err != nil || renamer == nil {`

I added a comment above to try to explain the line (`// Write to file and exit if not renaming`), but let me know if you have any suggestions on how to make it more clear.